### PR TITLE
Blob: Add types

### PR DESCRIPTION
**Pull request by sirreal**
## Description

Part of #18838
Extracted from #18942 

Add types to `blob` package.

**Testing Instructions:**

Verify type checking passes:

```
npm run lint-types
```

### CODE REVIEW

File: packages/blob/src/index.js [hunk 1]
```diff
@@ -3,6 +3,9 @@
  */
 const { createObjectURL, revokeObjectURL } = window.URL;
 
+/**
+ * @type {{[key: string]: File|undefined}}
```
Comment (sirreal) [inline packages/blob/src/index.js]: Initially, my editor ignored the `| undefined` part of this type and the following one. Adding explicit `strictNullChecks` to the tsconfig restored it.

File: packages/blob/src/index.js [hunk 2]
```diff
@@ -27,7 +30,7 @@ export function createBlobURL( file ) {
  *
  * @param {string} url The blob URL.
  *
- * @return {?File} The file for the blob URL.
+ * @return {File|undefined} The file for the blob URL.
```
Comment (sirreal) [inline packages/blob/src/index.js]: `{?File}` is a _nullable_ type in JSDoc, equivalent to `File | null`, which isn't accurate.

File: tsconfig.json [hunk 3]
```diff
@@ -12,6 +12,7 @@
 
 		/* Strict Type-Checking Options */
 		"strict": true,                        /* Enable all strict type-checking options. */
+		"strictNullChecks": true,
```
Comment (sirreal) [inline tsconfig.json]: This should be enabled via `strict`, but I some `Type | undefined` types didn't seem to be handled correctly. Adding this seems to fix the issue.

I wonder if `strict` + `noImplicitAny` results in odd behavior.

File: packages/blob/src/index.js (hunk 2 above)
Comment (aduth) [reply packages/blob/src/index.js]: > `{?File}` is a _nullable_ type in JSDoc, equivalent to `File | null`, which isn't accurate.

Yeah, I think you'll find a lot of examples like this in the codebase, where nullable type is used incorrectly as interchangeable with "optional" or "possibly undefined".

**Review by aduth — APPROVED** [review]

**Pull request merged**